src/main/java/org/folio/innreach/specification/InnReachTransactionSpecification.java

```diff
@@ -0,0 +1,99 @@
+package org.folio.innreach.specification;
+
+import org.folio.innreach.domain.entity.InnReachTransactionSortingParameters;
+import org.folio.innreach.domain.entity.InnReachTransaction;
+import org.folio.innreach.domain.entity.TransactionItemHold;
+import org.folio.innreach.domain.entity.TransactionLocalHold;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+
+import static org.folio.innreach.domain.entity.InnReachTransaction.TransactionType.ITEM;
+import static org.folio.innreach.domain.entity.InnReachTransaction.TransactionType.LOCAL;
+
+@Component
+public class InnReachTransactionSpecification {
+
+  public Specification<InnReachTransaction> filterByParameters(InnReachTransactionSortingParameters parameters) {
+    return isOfType(parameters.getTypes())
+      .and(isOfState(parameters.getStates()))
+      .and(centralCodeIn(parameters.getCentralServerCodes()))
+      .and(patronAgencyIn(parameters.getPatronAgencyCodes()))
+      .and(itemAgencyIn(parameters.getItemAgencyCodes()))
+      .and(patronTypeIn(parameters.getPatronTypes()))
+      .and(centralItemTypeIn(parameters.getCentralItemTypes()));
+  }
+
+  static Specification<InnReachTransaction> isOfType(List<InnReachTransaction.TransactionType> types) {
+    return (transaction, cq, cb) -> {
+      if (Long.class != cq.getResultType()) {
+        var hold = transaction.fetch("hold");
+        hold.fetch("pickupLocation");
+      }
+      if (types == null || types.isEmpty()) {
```

Comment:
"if" statement and the following return can be simplified to:
`return CollectionUtils.isEmpty(types) ? cb.conjunction() : transaction.get("type").in(types);`

the same is true for similar cases in the other methods